### DISCUSSION TITLE
Using Ipopt with Visual C++

### DISCUSSION THREAD

**Discussion by bradbell** [coin-or/Ipopt#588] [Q&A]
1. I know that Ipopt has a msys build, but does it have a Visual C++ build ?
2. If so does it generate a pkg-config file for use with such a build ?
3. If so how does on get a version of pkg-config that works for such a build ?

**Comment by svigerske** [comment]
Ipopt has an autotools based buildsystem, which can also be used under MSys2 to build with MS compilers.    
There is (no more) a MSVS buildsystem. I am not going to develop and maintain more than one buildsystem and as the current one works sufficiently well (in my opinion), I'm not going to change it.

When building Ipopt (with its one and only buildsystem), you get a pkg-config file installed. (They are also included in the libraries for download.)  For example,
```
prefix=${pcfiledir}/../..
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include/coin-or

Name: Ipopt
Description: Interior Point Optimizer
URL: https://github.com/coin-or/Ipopt
Version: 3.14.7
Cflags: -I${includedir}
Libs: -L${libdir} -lipopt
Requires.private: coinmumps 
```
These files should be fine to use on Unix-like systems and in an autotools-based build under Windows/MSys2. They are probably not directly usable in MSVS, because flags like `-L` and `-l` aren't understood there and paths are in Unix-style. I think I heard about versions of pkg-config that can translate between the GCC-style linker flags and the ones expected by the MS compilers, though. I cannot point you to such a version, sorry.

**Reply by tkralphs** [reply]
There is an open issue in BuildTools to emit pkg-config files with Windows-style paths: https://github.com/coin-or-tools/BuildTools/issues/132.

**Reply by tkralphs** [reply]
> So you are suggesting building ipopt using msys2 and then using that in a Visual C++ program build. There would still be the issue of how to use pkg-config in a Visual Studio development shell. I think that conda provides such a pkg-config, but I have not seen it work.

Yes. I don't know why you would need `pkg-config`. Just put the library somewhere that Visual Studio can find it and list it as a dependency. That's how all our Visual Studio solutions work.

**Reply by bradbell** [reply]
> Yes. I don't know why you would need `pkg-config`. Just put the library somewhere that Visual Studio can find it and list it as a dependency. That's how all our Visual Studio solutions work.

cmake uses pkg-config so that the same install script works no matter what system one is installing to.

**Reply by tkralphs** [reply]
Ah, so you are not actually building using Visual Studio, just using the Visual Studio compiler from CMake. Perhaps you could just do the path conversion yourself: https://stackoverflow.com/questions/29397349/cmake-convert-unix-to-windows-path.

**Reply by bradbell** [reply]
I am testing using CppAD with Ipopt when compiling with Visual C++. CMake is building the tests and the easy way to do this is to call pkgconfig from cmake and get the information for how Ipopt was built (using Visual C++).

**Reply by tkralphs** [reply]
Right. As I said, in that case, I would assume you can do the path conversion yourself.

**Comment by bradbell** [comment]
So you are suggesting building ipopt using msys2 and then using that in a Visual C++ program build. There would still be the issue of how to use pkg-config in a Visual Studio development shell. I think that conda provides such a pkg-config, but I have not seen it work.
